### ISSUE TITLE
Implement Auto Differentiation

### ISSUE THREAD

**Issue by ssshathi** [ssshathi/PET-kaggle#3]
Implement Reverse Mode Auto Differentiation:
If we are able to understand how backpropagation works in the specific case of a loss function,
we should consider implementing reverse mode auto differentiation: the generalization of backpropagation.

Useful Link:
[Stanford's backprop tutorial](http://cs231n.github.io/optimization-2/)

The issue is considered closed when:

- [ ] A concept of a computational graph is introduced (Make more detailed issue)
- [ ] Parameters are added as nodes to the graph, and operators as edges (Make more detailed issue)
- [ ] Auto Diff produces same result as previously written backprop algorithms
